### DISCUSSION TITLE
How to work with InlineKeyboardMarkup?

### DISCUSSION THREAD

**Discussion by MLNW** [visciang/telegram#149] [Q&A] [answered]
I would like to use [InlineKeyboardMarkup](https://core.telegram.org/bots/api#inlinekeyboardmarkup) in my bot to implement a configuration menu similar to what is used by the @BotFather for configuring bots.

From the [Telegram API docs](https://core.telegram.org/bots/api#sendmessage) it sounds to me like I just need to send a message with a `reply_markup` field set to a JSON array of array of [InlineKeyboardButton](https://core.telegram.org/bots/api#inlinekeyboardbutton)s.

I tried it with code like this:
```elixir
keyboard =
  [
    [%{text: "Configure A", callback_data: "a_start"}],
    [%{text: "Configure B", callback_data: "b_start"}]
  ]
  |> Jason.encode!()

Telegram.Api.request(token, "sendMessage",
  text: "Configuration Menu",
  chat_id: chat_id,
  reply_markup: keyboard)
```

The result is sadly nothing at all. As soon as I add the `reply_markup` parameter the API call seems to fail silently. I had a similar issue before when trying to send a message with `parse_mode: "MarkdownV2"` that included a `.` in the message text.

1. Is there a way to turn on additional logging or is there another way to figure out what is happening when I run this?
2. Do you have an idea what I'm doing wrong?

I'm fairly new to Elixir, so it might just be a stupid mistake on my side.

**Comment by visciang** [comment, marked as answer]
Hi @MLNW 

Have a look here https://visciang.github.io/telegram/readme.html#sendmessage-with-keyboard
Just use the json object "facility" (no need to serialize the object yourself) and use the correct `reply_markup` object schema (`keyboard_markup = %{one_time_keyboard: ..., keyboard: ...}`)

About errors: we are just send request to the telegram server and you get back a reply from them, just look at the response.
But consider that they will apply the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle), so if you send by mistake extra info that won't give any error.

IE:

```elixir
iex(5)> Telegram.Api.request(token, "sendMessage", chat_id: 0, text: "text")
{:error, "Bad Request: chat not found"}

iex(6)> Telegram.Api.request(token, "getMe", this_is_an_unknown_parameter: 1)
{:ok, %{...}}
```

Hope it helps you

**Reply by MLNW** [reply]
Thank you very much for the quick and great reply. As expected it comes down to mistakes on my part. Somehow I even missed the mention of the keyboard in your docs even though I must have read through it multiple times.

I'll try it out tonight and let you know how it goes.

**Reply by MLNW** [reply]
Yes, this is exaclty what I was looking for and it works well! Generally, I have gained a better understanding of how your lib works with the Telegram API. Thanks for that!

I'm now going to build a configuration menu around the `inline_keyboard`.
